download_service/zipbuilder.py

```diff
@@ -0,0 +1,126 @@
+from zipstream import ZipFile
+from ddsc.sdk.client import PathToFiles
+from ddsc.core.ddsapi import DataServiceError
+import requests
+
+
+class ZipBuilderException(BaseException):
+    def __init__(self, message):
+        self.message = message
+
+
+class NotFoundException(ZipBuilderException):
+    pass
+
+
+class NotSupportedException(ZipBuilderException):
+    pass
+
+
+class DDSZipBuilder(object):
+    """
+    Builds a zip file as a stream, containing all the files in a DukeDS project.
+    Notes about generators and timeliness
+    """
+
+    def __init__(self, project_id, client):
+        """
+
+        :param project_id: The id of a DukeDS project
+        :param client: A ddsc.sdk.Client instance ready to make API calls
+        """
+        self.project_id = project_id
+        self.client = client
+
+    @staticmethod
+    def _handle_dataservice_error(e, message):
+        if e.status_code == 404:
+            raise NotFoundException(message)
+        else:
+            raise
+
+    def get_project_name(self):
+        """
+        Uses the client to look up the project name from DukeDS
+        :return: str: project name
+        """
+        try:
+            project = self.client.get_project_by_id(self.project_id)
+            return project.name
+        except DataServiceError as e:
+            DDSZipBuilder._handle_dataservice_error(e, 'Project {} not found'.format(self.project_id))
+
+    def get_filename(self):
+        """
+        Generates a file name for the zip file based on the project name
+        :return: str: file name ending in .zip
+        """
+        project_name = self.get_project_name()
+        return '{}.zip'.format(project_name)
+
+    def get_dds_paths(self):
+        """
+        Builds a mapping of file paths (strings) in the dds project to ddsc.sdk.client.File objects
+        :return: OrderedDict where keys are relative paths in the project and values are ddsc.sdk.client.File objects
+        """
+        try:
+            children = self.client.dds_connection.get_project_children(self.project_id)
+            ptf = PathToFiles()
+            for child in children:
+                ptf.add_paths_for_children_of_node(child)
+            return ptf.paths  # OrderedDict of path -> File
```

Comment:
I got the following error when trying to download a project with top level files:
```
...
  File "/Users/jpb67/Documents/work/D4S2/download_service/zipbuilder.py", line 70, in get_dds_paths
    ptf.add_paths_for_children_of_node(child)
  File "/Users/jpb67/Documents/work/D4S2/venv/lib/python3.6/site-packages/ddsc/sdk/client.py", line 513, in add_paths_for_children_of_node
    self._child_recurse(node, '')
  File "/Users/jpb67/Documents/work/D4S2/venv/lib/python3.6/site-packages/ddsc/sdk/client.py", line 516, in _child_recurse
    for child in node.get_children():
  File "/Users/jpb67/Documents/work/D4S2/venv/lib/python3.6/site-packages/ddsc/sdk/client.py", line 260, in __getattr__
    raise AttributeError(msg)
AttributeError: 'File' object has no attribute 'get_children'
```
I suggest replacing the contents of this try/except above with:
```
            project = self.client.get_project_by_id(self.project_id)
            ptf = PathToFiles()
            ptf.add_paths_for_children_of_node(project)
            return ptf.paths
```

Reply:
Great catch, thanks!